--- a/src/core/arm/dynarmic/arm_dynarmic_64.cpp
+++ b/src/core/arm/dynarmic/arm_dynarmic_64.cpp
@@ -5,6 +5,7 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 
 #include "common/settings.h"
+#include "core/arm/debug.h"
 #include "core/arm/dynarmic/arm_dynarmic.h"
 #include "core/arm/dynarmic/arm_dynarmic_64.h"
 #include "core/arm/dynarmic/dynarmic_exclusive_monitor.h"
@@ -107,13 +108,13 @@
     case Dynarmic::A64::InstructionCacheOperation::InvalidateByVAToPoU: {
         static constexpr u64 ICACHE_LINE_SIZE = 64;
         const u64 cache_line_start = value & ~(ICACHE_LINE_SIZE - 1);
-        m_parent.InvalidateCacheRange(cache_line_start, ICACHE_LINE_SIZE);
+        Core::InvalidateInstructionCacheRange(m_process, cache_line_start, ICACHE_LINE_SIZE);
         break;
     }
     case Dynarmic::A64::InstructionCacheOperation::InvalidateAllToPoU:
+    case Dynarmic::A64::InstructionCacheOperation::InvalidateAllToPoUInnerSharable:
         m_parent.ClearInstructionCache();
         break;
-    case Dynarmic::A64::InstructionCacheOperation::InvalidateAllToPoUInnerSharable:
     default:
         LOG_DEBUG(Core_ARM, "Unprocesseed instruction cache operation: {}", op);
         break;
@@ -458,6 +459,7 @@
 }
 
 void ArmDynarmic64::InvalidateCacheRange(u64 addr, std::size_t size) {
+    m_cb->last_code_addr = u64(-1);
     m_jit->InvalidateCacheRange(addr, size);
 }
 
--- a/src/core/arm/dynarmic/arm_dynarmic_32.cpp
+++ b/src/core/arm/dynarmic/arm_dynarmic_32.cpp
@@ -432,6 +432,7 @@
 }
 
 void ArmDynarmic32::InvalidateCacheRange(u64 addr, std::size_t size) {
+    m_cb->last_code_addr = u64(-1);
     m_jit->InvalidateCacheRange(static_cast<u32>(addr), size);
 }
 
