--- a/src/common/net/net.cpp
+++ b/src/common/net/net.cpp
@@ -226,12 +226,15 @@
             .method = "GET",
             .path = path,
         };
+        request.headers.emplace("User-Agent", "Prospero.Eden-Encore/1");
+        request.headers.emplace("Accept", "*/*");
 
         client->set_follow_location(true);
         httplib::Result result = client->send(request);
 
         if (!result) {
-            LOG_ERROR(Common, "GET to {}{} returned null", url, path);
+            LOG_ERROR(Common, "GET to {}{} returned null: {}", url, path,
+                      httplib::to_string(result.error()));
             return {};
         }
 
